homeassistant/components/mqtt/humidifier.py

```diff
@@ -284,164 +280,166 @@ def add_subscription(
-                return
-            if (
-                target_humidity < self._attr_min_humidity
-                or target_humidity > self._attr_max_humidity
-            ):
-                _LOGGER.warning(
-                    "'%s' received on topic %s. '%s' is not a valid target humidity",
-                    msg.payload,
-                    msg.topic,
-                    rendered_target_humidity_payload,
-                )
-                return
-            self._attr_target_humidity = target_humidity
+            return
+        if (
+            target_humidity < self._attr_min_humidity
+            or target_humidity > self._attr_max_humidity
+        ):
+            _LOGGER.warning(
+                "'%s' received on topic %s. '%s' is not a valid target humidity",
+                msg.payload,
+                msg.topic,
+                rendered_target_humidity_payload,
+            )
+            return
+        self._attr_target_humidity = target_humidity
 
-        self.add_subscription(
-            topics, CONF_TARGET_HUMIDITY_STATE_TOPIC, target_humidity_received
-        )
+    @callback
+    def _mode_received(self, msg: ReceiveMessage) -> None:
+        """Handle new received MQTT message for mode."""
+        mode = str(self._value_templates[ATTR_MODE](msg.payload))
+        if mode == self._payload["MODE_RESET"]:
+            self._attr_mode = None
+            return
+        if not mode:
+            _LOGGER.debug("Ignoring empty mode from '%s'", msg.topic)
+            return
+        if not self.available_modes or mode not in self.available_modes:
+            _LOGGER.warning(
+                "'%s' received on topic %s. '%s' is not a valid mode",
+                msg.payload,
+                msg.topic,
+                mode,
+            )
+            return
 
-        @callback
-        @log_messages(self.hass, self.entity_id)
-        @write_state_on_attr_change(self, {"_attr_mode"})
-        def mode_received(msg: ReceiveMessage) -> None:
-            """Handle new received MQTT message for mode."""
-            mode = str(self._value_templates[ATTR_MODE](msg.payload))
-            if mode == self._payload["MODE_RESET"]:
-                self._attr_mode = None
-                return
-            if not mode:
-                _LOGGER.debug("Ignoring empty mode from '%s'", msg.topic)
-                return
-            if not self.available_modes or mode not in self.available_modes:
-                _LOGGER.warning(
-                    "'%s' received on topic %s. '%s' is not a valid mode",
-                    msg.payload,
-                    msg.topic,
-                    mode,
-                )
-                return
+        self._attr_mode = mode
 
-            self._attr_mode = mode
+    def _prepare_subscribe_topics(self) -> None:
+        """(Re)Subscribe to topics."""
+        topics: dict[str, Any] = {}
 
-        self.add_subscription(topics, CONF_MODE_STATE_TOPIC, mode_received)
```

Comment:
This pattern is nicer than fan since it avoids the inner function here

Reply:
We can refactor this later, I did not want to change too much.